controllers/imagehub/api/v1/image_webhook.go

```diff
@@ -17,69 +17,95 @@ limitations under the License.
 package v1
 
 import (
+	"context"
+	"encoding/json"
 	"fmt"
+	"net/http"
 
-	"k8s.io/apimachinery/pkg/runtime"
-	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
-	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 )
 
 // log is for logging in this package.
 var imagelog = logf.Log.WithName("image-resource")
 
-func (i *Image) SetupWebhookWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewWebhookManagedBy(mgr).
-		For(i).
-		Complete()
-}
-
-// TODO(user): EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
-
 //+kubebuilder:webhook:path=/mutate-imagehub-sealos-io-v1-image,mutating=true,failurePolicy=fail,sideEffects=None,groups=imagehub.sealos.io,resources=images,verbs=create;update,versions=v1,name=mimage.kb.io,admissionReviewVersions=v1
+//+kubebuilder:object:generate=false
 
-var _ webhook.Defaulter = &Image{}
+// ImageMutater add lables to Images
+type ImageMutater struct {
+	Client  client.Client
+	decoder *admission.Decoder
+}
 
-// Default implements webhook.Defaulter so a webhook will be registered for the type
-func (i *Image) Default() {
+func (m *ImageMutater) Handle(ctx context.Context, req admission.Request) admission.Response {
```

Comment:
I'm aware of it, but I need req.Userinfo for Validator. This is the only to get it? Defaulter can change to this, I'm not sure that it will run properly in both two ways(use defaulter and registe my Validator at the same time).

Reply:
This is where I got this codes.[kubebuilder.io](https://book.kubebuilder.io/reference/webhook-for-core-types.html)

Reply:
> I'm aware of it, but I need req.Userinfo for Validator.Is this the only to get it?
> 
> Defaulter can change to this, but I'm not sure that it will run properly in both two ways(use defaulter and registe my Validator at the same time).

```
req, err := admission.RequestFromContext(ctx)
	if err != nil {
		return err
	}
```
This can get all the information you want.

Reply:
It works, thx a lot.

Reply:
I have solved it. see this:https://github.com/lingdie/sealos/blob/3185f0128b5ff62220da5377883515349403dc6f/controllers/imagehub/api/v1/image_webhook.go#L34